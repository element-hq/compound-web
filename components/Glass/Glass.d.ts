import React, { ComponentPropsWithoutRef, ReactNode } from "react";
interface Props extends ComponentPropsWithoutRef<"div"> {
    /**
     * The child component.
     */
    children: ReactNode;
    /**
     * The CSS class.
     */
    className?: string;
}
/**
 * Adds a border of glass around a child component.
 */
export declare const Glass: React.ForwardRefExoticComponent<Props & React.RefAttributes<HTMLDivElement>>;
export {};
//# sourceMappingURL=Glass.d.ts.map