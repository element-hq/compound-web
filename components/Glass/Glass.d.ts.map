{"version":3,"file":"Glass.d.ts","sourceRoot":"","sources":["../../../src/components/Glass/Glass.tsx"],"names":[],"mappings":"AAgBA,OAAO,KAAK,EAAE,EACZ,wBAAwB,EACxB,SAAS,EAGV,MAAM,OAAO,CAAC;AAKf,UAAU,KAAM,SAAQ,wBAAwB,CAAC,KAAK,CAAC;IACrD;;OAEG;IACH,QAAQ,EAAE,SAAS,CAAC;IACpB;;OAEG;IACH,SAAS,CAAC,EAAE,MAAM,CAAC;CACpB;AAED;;GAEG;AACH,eAAO,MAAM,KAAK,8EAMjB,CAAC"}