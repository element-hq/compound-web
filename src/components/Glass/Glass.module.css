/*
Copyright 2023 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

.glass {
  border-radius: var(--cpd-space-9x);
  padding: var(--cpd-space-3x);

  /* We use an outline here to create an "inner border", rather than one that
  adds to the component's size */
  outline: var(--cpd-border-width-1) solid var(--cpd-color-alpha-gray-400);
  outline-offset: calc(-1 * var(--cpd-border-width-1));
  background: var(--cpd-color-alpha-gray-400);
  backdrop-filter: blur(20px);
}

.glass > :first-child {
  border-radius: var(--cpd-space-6x);
  inline-size: 100%;
  block-size: 100%;
}
