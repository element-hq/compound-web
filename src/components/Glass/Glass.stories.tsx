/*
Copyright 2023 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import React from "react";
import { Meta, StoryObj } from "@storybook/react-vite";

import { Glass as GlassComponent } from "./Glass";
import { Text } from "../Typography/Text";

const meta = {
  title: "Glass",
  component: GlassComponent,
  tags: ["autodocs"],
  argTypes: {},
  args: {
    style: {
      inlineSize: 250,
      blockSize: 150,
    },

    children: (
      <div
        style={{
          background: "var(--cpd-color-bg-canvas-default)",
          display: "flex",
          alignItems: "center",
          justifyContent: "center",
        }}
      >
        <Text style={{ textAlign: "center" }}>Your content here</Text>
      </div>
    ),
  },

  decorators: [
    (Story) => (
      <div
        style={{
          background: "url('/images/__test__/kitten2.jpg')",
          backgroundSize: "cover",
          inlineSize: "fit-content",
          blockSize: "fit-content",
          padding: 16,
        }}
      >
        <Story />
      </div>
    ),
  ],
} satisfies Meta<typeof GlassComponent>;
export default meta;

type Story = StoryObj<typeof meta>;

export const Glass: Story = { args: {} };
