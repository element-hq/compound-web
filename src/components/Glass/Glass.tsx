/*
Copyright 2023 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import React, {
  type ComponentPropsWithoutRef,
  type ReactNode,
  forwardRef,
  Children,
} from "react";
import classNames from "classnames";

import styles from "./Glass.module.css";

interface Props extends ComponentPropsWithoutRef<"div"> {
  /**
   * The child component.
   */
  children: ReactNode;
  /**
   * The CSS class.
   */
  className?: string;
}

/**
 * Adds a border of glass around a child component.
 */
export const Glass = forwardRef<HTMLDivElement, Props>(
  ({ children, className, ...rest }, ref) => (
    <div ref={ref} className={classNames(className, styles.glass)} {...rest}>
      {Children.only(children)}
    </div>
  ),
);

Glass.displayName = "Glass";
