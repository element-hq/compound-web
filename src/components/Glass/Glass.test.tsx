/*
Copyright 2023 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import { describe, it, expect } from "vitest";
import { render } from "@testing-library/react";
import React from "react";

import { Glass } from "./Glass";

describe("Glass", () => {
  it("renders", () => {
    const { asFragment } = render(
      <Glass>
        <div>ooh, shinyâ€¦</div>
      </Glass>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
