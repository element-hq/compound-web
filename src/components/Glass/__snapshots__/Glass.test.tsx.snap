// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Glass > renders 1`] = `
<DocumentFragment>
  <div
    class="_glass_77c50d"
  >
    <div>
      ooh, shinyâ€¦
    </div>
  </div>
</DocumentFragment>
`;
